Fixes #35167 - move hostgroups dropdown actions to extendable helper (#9297)
theforeman/foreman: developer_docs/how_to_create_a_plugin.asciidoc

----

[[extending-hostgroup-ui]]
=== Extending hostgroup UI

> A plugin can add items displayed in the `Actions` dropdown on hostgroup overview page.

Reviewer: ```suggestion
A plugin can add items to the `Actions` dropdown in the table on the hostgroups overview page.
```

Adding an item to the actions dropdown requires adding a helper with a method named `hostgroups_dropdown_actions` that accepts a hostgroup as an argument

> and returns an array of relevant actions to your plugin.

Reviewer: can you give an example of how you extended it from foreman_ansible? https://github.com/theforeman/foreman_ansible/pull/541

just the code example to extend the `hostgroups_dropdown_actions` helper

Author: I added an example, let me know if it is ok, or if I should change it.

----

[[extending-hostgroup-ui]]
=== Extending hostgroup UI
A plugin can add items to the `Actions` dropdown in the table on the hostgroups overview page.

Adding an item to the actions dropdown requires adding a helper with a method named `hostgroups_dropdown_actions` that accepts a hostgroup as an argument
and returns an array of relevant actions to your plugin.

In plugin helper (my_plugin_helper.rb):
[source, ruby]
....
def hostgroups_dropdown_actions(hostgroup)
   new_item = link_to _('Link Title'), :other_properties

   # to add to the existing actions
   super.insert(new_item)

   # to override the existing actions

>    [new_item]

Reviewer: Actually, it feels a bit risky, as we give plugins the option to remove items from core or from other plugins...
while this is a legit solution, today we discussed with @ShimShtein on IRC a solution that I think fits better.

here is a similar implementation:
https://github.com/theforeman/foreman/commit/377ff667bbe3c2df8cd0831d9fe60bb9b316d680#diff-77bd7ad90032b71e085bbb163dfd6e5795c32dabf5de551a32b10d0cf837c87fR122

and from a plugin, could extend foreman this way: https://github.com/theforeman/foreman/blob/develop/developer_docs/how_to_create_a_plugin.asciidoc#adding-overview-details-buttons

----

[[extending-hostgroup-ui]]
=== Extending hostgroup UI

> A plugin can add items to the `Actions` dropdown in the table on the hostgroups overview page.

Reviewer 2: I suggest adding the description of the action to the TOC:
```suggestion
[[extending-hostgroup-ui-index-actions]]
==== Adding actions to the index page
A plugin can add items to the `Actions` dropdown in the table on the hostgroups overview page.
```

----

[[extending-hostgroup-ui]]
=== Extending hostgroup UI

[[extending-hostgroup-ui-index-actions]]
==== Adding actions to the index page
A plugin can add items to the `Actions` dropdown in the table on the hostgroups overview page.

Adding an item to the actions dropdown requires adding a helper with a method that accepts a hostgroup as an argument and
returns a list of hashes, where each hash will
have two predefined fields: `:action` and `:priority`. The `:priority` value would be used by the
system to define the order of the items to show. The lower the priority, the
higher the item will show.

In plugin helper (`my_plugin_helper.rb`):
[source, ruby]
----
def my_plugin_hostgroups_actions(hostgroup)
  [

>     { :action => [_('new_action'), other_properties], :priority => 20 }

Reviewer 2: ```suggestion
    { :action => display_link_if_authorized('foo', {bar: 'bar'}, {baz: 'baz'}), :priority => 20 }
```
The item should be an HTML element (probably link) that will be embedded as an item in the dropdown

Author: You are right, I need to update the documentation. Thanks!
